### PR TITLE
implemented mesh traversal for validation of active set 

### PR DESCRIPTION
**Pull request by antonlerner**
implemented mesh traversal for validation of active set state in atxs

I've refactored some files into mesh so just skim over activaiton.go and atxdb.go

new code is in mesh

### CODE REVIEW

File: consensus/ninja_tortoise.go [hunk 1]
```diff
@@ -64,7 +64,7 @@ func (vp votingPattern) Layer() mesh.LayerID {
 type BlockCache interface {
 	GetBlock(id mesh.BlockID) (*mesh.Block, error)
 	LayerBlockIds(id mesh.LayerID) ([]mesh.BlockID, error)
-	ForBlockInView(view map[mesh.BlockID]struct{}, layer mesh.LayerID, foo func(block *mesh.BlockHeader), errHandler func(err error))
+	ForBlockInView(view map[mesh.BlockID]struct{}, layer mesh.LayerID, foo mesh.TraversalFunc, errHandler func(err error)) error
```
Comment (noamnelke) [inline consensus/ninja_tortoise.go]: Let's rename `foo`?

File: consensus/ninja_tortoise.go [hunk 2]
```diff
@@ -211,7 +211,7 @@ func globalOpinion(v vec, layerSize uint64, delta float64) vec {
 }
 
 func (ni *ninjaTortoise) updateCorrectionVectors(p votingPattern, bottomOfWindow mesh.LayerID) {
-	foo := func(x *mesh.BlockHeader) {
+	foo := func(x *mesh.BlockHeader) error {
```
Comment (noamnelke) [inline consensus/ninja_tortoise.go]: `foo` and `x` aren't great names...

File: consensus/ninja_tortoise.go [hunk 3]
```diff
@@ -462,13 +463,14 @@ func (ni *ninjaTortoise) handleIncomingLayer(newlyr *mesh.Layer) { //i most rece
 			view := make(map[mesh.BlockID]struct{})
 			lCntr := make(map[mesh.LayerID]int)
 			correctionMap, effCountMap, getCrrEffCnt := ni.getCorrEffCounter()
-			foo := func(block *mesh.BlockHeader) {
+			foo := func(block *mesh.BlockHeader) error {
```
Comment (noamnelke) [inline consensus/ninja_tortoise.go]: same here

File: mesh/activation.go [hunk 4]
```diff
@@ -101,3 +127,106 @@ func BytesAsAtx(b []byte) (*ActivationTx, error) {
...
+	epochProvider EpochProvider
+}
+
+func NewBuilder(nodeId NodeId, db database.DB, net Broadcaster, activeSet ActiveSetProvider, view MeshProvider, epochDuration EpochProvider) *Builder {
+	return &Builder{
+		nodeId, NewActivationDb(db), net, activeSet, view, epochDuration,
+	}
+}
+
+func (b *Builder) PublishActivationTx(nipst *nipst.NIPST) error {
```
Comment (noamnelke) [inline mesh/activation.go]: I would change the way this method is organized:
- get prev atx id
- if not found: seq = 0 and prev atx id = nil
- else: get atx seq (passing the id into the method, so it doesn't need to find it again)

File: mesh/activation.go [hunk 5]
```diff
@@ -101,3 +127,106 @@ func BytesAsAtx(b []byte) (*ActivationTx, error) {
...
+	} else {
+		posAtx = &EmptyAtx
+	}
+
+	if prevAtx == nil {
+		log.Info("previous ATX not found")
+		prevAtx = &EmptyAtx
+	}
+
+	atx := NewActivationTx(b.nodeId, seq+1, *prevAtx, l, 0, *posAtx, b.activeSet.GetActiveSetSize(l-1), b.mesh.GetLatestView(), nipst)
```
Comment (noamnelke) [inline mesh/activation.go]: `seq+1` means counting starts at 1 instead of 0.

If we change the organization as I suggested we can do the incrementation only if a prev atx was found.

File: mesh/atxdb.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = db.addAtxToEpoch(ech, atx.Id())
+	if err != nil {
+		return err
+	}
+	if atx.Valid() {
+		db.incValidAtxCounter(ech)
```
Comment (noamnelke) [inline mesh/atxdb.go]: unhandled error

File: mesh/atxdb.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		return err
+	}
+	return nil
+}
+
+func epochCounterKey(ech EpochId) []byte {
+	return append(ech.ToBytes(), common.Uint64ToBytes(uint64(CounterKey))...)
+}
+
+func (db *ActivationDb) incValidAtxCounter(ech EpochId) error {
```
Comment (noamnelke) [inline mesh/atxdb.go]: Can this never run concurrently? What if get multiple activations around the same time (sounds likely to me)?

File: consensus/ninja_tortoise.go (hunk 1 above)
Comment (almogdepaz) [inline consensus/ninja_tortoise.go]: i prefer it stayed func(block *mesh.BlockHeader) or whatever 
it dosen't have to be related to mesh

File: mesh/atxdb.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return atx, nil
+}
+
+func decodeAtxIds(idsBytes []byte) ([]AtxId, error) {
+	var ids []AtxId
+	if _, err := xdr.Unmarshal(bytes.NewReader(idsBytes), &ids); err != nil {
+		return nil, errors.New("error marshaling layer ")
```
Comment (noamnelke) [inline mesh/atxdb.go]: you probably mean `unmarshaling`...

Don't we want to include the actual error and also log it?

File: mesh/atxdb.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+	ids, err := db.atxs.Get(epochId.ToBytes())
+	var atxs []AtxId
+	if err != nil {
+		//epoch doesnt exist, need to insert new layer
+		ids = []byte{}
+		atxs = make([]AtxId, 0, 0)
+	} else {
+		atxs, err = decodeAtxIds(ids)
+		if err != nil {
+			return errors.New("could not get all atxs from database ")
```
Comment (noamnelke) [inline mesh/atxdb.go]: I think we should include the previous error and log this error

File: mesh/activation.go [hunk 10]
```diff
@@ -6,12 +6,17 @@ import (
...
+	"github.com/spacemeshos/go-spacemesh/log"
 	"github.com/spacemeshos/go-spacemesh/nipst"
+	"github.com/spacemeshos/go-spacemesh/rand"
 )
 
 //todo: choose which type is VRF
 type Vrf string
 
+const AtxProtocol = "AtxGossip"
+
```
Comment (almogdepaz) [inline mesh/activation.go]: why is this here?, i think mesh should be kept separated from anything related to network i.e broadcasting listening ...

File: mesh/activation.go [hunk 11]
```diff
@@ -101,3 +127,106 @@ func BytesAsAtx(b []byte) (*ActivationTx, error) {
...
+	LatestLayerId() LayerID
+}
+
+type EpochProvider interface {
+	Epoch(l LayerID) EpochId
+}
+
+type Broadcaster interface {
+	Broadcast(channel string, data []byte) error
+}
```
Comment (almogdepaz) [inline mesh/activation.go]: like i said im not sure this is a good idea
i think it should be in a separate package 
if the types are an issue maybe its time to pull them out of here as well

File: p2p/dht/table.go [hunk 12]
```diff
@@ -109,7 +109,7 @@ type routingTableImpl struct {
 	// Local peer ID that holds this routing table
 	local node.DhtID
 
-	// Operations activation channels
+	// Operations mesh channels
```
Comment (almogdepaz) [inline p2p/dht/table.go:112]: refactor error

File: mesh/mesh.go [hunk 13]
```diff
@@ -8,16 +8,17 @@ import (
...
 	"sync/atomic"
 )
 
 const (
-	layerSize = 200
-	Genesis   = 0
-	GenesisId = 420
+	layerSize     = 200
+	Genesis       = 0
+	GenesisId     = 420
```
Comment (noamnelke) [inline mesh/mesh.go]: `go fmt ./...`

File: mesh/mesh.go [hunk 14]
```diff
@@ -51,43 +58,44 @@ type Mesh struct {
 	done          chan struct{}
 }
 
-func NewPersistentMesh(path string, rewardConfig RewardConfig, mesh MeshValidator, state StateUpdater, logger log.Log) *Mesh {
+func NewPersistentMesh(path string, rewardConfig Config, mesh MeshValidator, state StateUpdater, logger log.Log) *Mesh {
```
Comment (noamnelke) [inline mesh/mesh.go]: rename the argument name as well?

File: mesh/mesh.go [hunk 15]
```diff
@@ -51,43 +58,44 @@ type Mesh struct {
...
+		done:     make(chan struct{}),
+		mdb:      NewPersistentMeshDB(path, logger),
+		config:   rewardConfig,
 	}
 
 	return ll
 }
 
-func NewMemMesh(rewardConfig RewardConfig, mesh MeshValidator, state StateUpdater, logger log.Log) *Mesh {
+func NewMemMesh(rewardConfig Config, mesh MeshValidator, state StateUpdater, logger log.Log) *Mesh {
```
Comment (noamnelke) [inline mesh/mesh.go]: same

File: mesh/mesh.go [hunk 16]
```diff
@@ -51,43 +58,44 @@ type Mesh struct {
...
+		done:     make(chan struct{}),
+		mdb:      NewMemMeshDB(logger),
+		config:   rewardConfig,
 	}
 
 	return ll
 }
 
-func NewMesh(db *MeshDB, rewardConfig RewardConfig, mesh MeshValidator, state StateUpdater, logger log.Log) *Mesh {
+func NewMesh(db *MeshDB, atxDb AtxDB, rewardConfig Config, mesh MeshValidator, state StateUpdater, logger log.Log) *Mesh {
```
Comment (noamnelke) [inline mesh/mesh.go:90]: same

File: mesh/reward.go [hunk 17]
```diff
@@ -6,25 +6,29 @@ import (
...
-type RewardConfig struct {
+type Config struct {
+	// reward config
 	SimpleTxCost   *big.Int
 	BaseReward     *big.Int
 	PenaltyPercent *big.Int
 	TxQuota        uint32
 	RewardMaturity LayerID
+	// global mesh config
+	LayersPerEpoch LayerID
```
Comment (noamnelke) [inline mesh/reward.go]: kind of weird that the type is `LayerID`

File: p2p/dht/table.go (hunk 12 above)
Comment (noamnelke) [inline p2p/dht/table.go:112]: intentional?

File: mesh/atxdb.go (hunk 7 above)
Comment (antonlerner) [reply mesh/atxdb.go]: yes you are correct, I'll make concurrency changes in another PR though, after this one is merged.

File: mesh/reward.go (hunk 17 above)
Comment (antonlerner) [reply mesh/reward.go]: removed

File: mesh/mesh.go (hunk 15 above)
Comment (antonlerner) [reply mesh/mesh.go]: I've removed config that is not reward related so it's not really relevant at the moment

File: activation/activation.go [hunk 18]
```diff
@@ -52,12 +52,21 @@ func NewBuilder(nodeId block.NodeId, db database.DB, meshdb *mesh.MeshDB, net Br
...
 	prevAtx, err := b.GetPrevAtxId(b.nodeId)
-	if seq > 0 && err != nil {
-		log.Error("cannot find prev ATX for nodeid %v ", b.nodeId)
-		return err
+	if prevAtx == nil {
+		log.Info("previous ATX not found")
+		prevAtx = &block.EmptyAtx
+		seq = 0
+	} else  {
+		seq = b.GetLastSequence(b.nodeId)
```
Comment (noamnelke) [inline activation/activation.go:61]: doesn't it make more sense to pass the `atxID` (and rename this `GetSequence`) instead of looking for the `prevAtx` again inside this method? Not a big deal, if you disagree.

File: activation/activation.go [hunk 19]
```diff
@@ -52,12 +52,21 @@ func NewBuilder(nodeId block.NodeId, db database.DB, meshdb *mesh.MeshDB, net Br
...
+		if seq > 0 && prevAtx == nil {
```
Comment (noamnelke) [inline activation/activation.go:62]: `prevAtx` can never be `nil` here.

**Review by noamnelke — APPROVED** [review]

**Review by almogdepaz — CHANGES_REQUESTED** [review]
i went over it and i dont think block is a good choise for the package name, i looked at go-ethereum and they named it core maybe we should go with something similar
or something more descriptive to the content of the package

**Comment by antonlerner** [comment]
> i went over it and i dont think block is a good choise for the package name, i looked at go-ethereum and they named it core maybe we should go with something similar
> or something more descriptive to the content of the package

 Core is also very generic and not descriptive, currently package block contains all data structures and types that exists in a spacemesh block. 
Refactoring the name of the package can be changed always but I cant think of more appropriate name. Your'e welcome to find something more descriptive but i don't want to spend more time on naming.

**Review by almogdepaz — APPROVED** [review]
